src/interfaces/simulation.ts

```diff
@@ -38,84 +44,145 @@ export class SimulationInterface<T extends ChainId> extends ServiceInterface<T>
-            })
-            .then((res) => res.simulation.id)
-        : undefined;
-      options.root = approvalTransactionId;
+    const { approvalTransactionId, forkId } = await this.getApprovalData({
+      depositProps: { from, sellToken, amount, toVault, options },
+      signer,
+    });
 
-      simulateDeposit = (save: boolean): Promise<TransactionOutcome> => {
-        options.save = save;
-        return this.zapIn(
-          from,
-          sellToken,
-          underlyingToken,
-          amount,
-          toVault,
-          vaultContract,
-          zapProtocol,
-          needsApproving,
-          options
-        );
-      };
-    } else {
-      const needsApproving = await this.depositNeedsApproving(from, sellToken, toVault, amount, signer);
+    const simulateFn = (save: boolean): Promise<TransactionOutcome> => {
+      return this.directDeposit({
+        depositProps: { from, sellToken, amount, toVault, options },
+        vaultContract,
+        options: { ...options, forkId, save, root: approvalTransactionId },
+      });
+    };
 
-      forkId = needsApproving ? await this.simulationExecutor.createFork() : undefined;
-      options.forkId = forkId;
+    return this.simulationExecutor.executeSimulationWithReSimulationOnFailure(simulateFn, forkId);
+  }
 
-      const approvalTransactionId = needsApproving
-        ? await this.approve(from, sellToken, amount, toVault, options)
-        : undefined;
-      options.root = approvalTransactionId;
+  private async getUnderlyingToken(
+    vaultContract: PickleJarContract | YearnVaultContract
+  ): Promise<{ address: Address; decimals: BigNumber; pricePerShare: BigNumber }> {
+    const [address, decimals, pricePerShare] = await Promise.all([
+      vaultContract.token().catch(() => {
+        throw new EthersError("failed to fetch token", EthersError.FAIL_TOKEN_FETCH);
+      }),
+      vaultContract.decimals().catch(() => {
+        throw new EthersError("no decimals", EthersError.NO_DECIMALS);
+      }),
+      vaultContract.pricePerShare().catch(() => {
+        throw new EthersError("no price per share", EthersError.NO_PRICE_PER_SHARE);
+      }),
+    ]);
 
-      simulateDeposit = (save: boolean): Promise<TransactionOutcome> => {
-        options.save = save;
-        return this.directDeposit(from, sellToken, amount, toVault, vaultContract, options);
-      };
+    return { address, decimals, pricePerShare };
+  }
+
+  private async directDeposit({
+    depositProps: { toVault, amount, from, sellToken },
+    vaultContract,
+    options,
+  }: {
+    depositProps: DepositProps;
+    vaultContract: VaultContract;
+    options: SimulationOptions;
+  }): Promise<TransactionOutcome> {
+    const encodedInputData = await (this.shouldUsePartnerService(toVault)
+      ? this.yearn.services.partner?.encodeDeposit(toVault, amount)
+      : vaultContract.encodeDeposit(amount));
+
+    if (!encodedInputData) {
+      throw new Error("directDeposit#encodeDeposit failed");
```

Comment:
This should be an SdkError right?

Reply:
I didn't change this, just moved it around. Can refactor, we have about 33 `new Error` in the SDK, should they all be `SdkError`?

Reply:
they probably should, but if this is just from moving around no need to do it now